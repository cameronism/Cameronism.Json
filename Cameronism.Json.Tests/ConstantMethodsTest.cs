using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace Cameronism.Json.Tests
{
	public class ConstantMethodsTest
	{
		#region properties
		string C2 { get { return null; } }
		string C3 { get { return "a"; } }
		int C4 { get { return 1; } }
		int? C5 { get { return 1; } }
		int? C6 { get { return null; } }
		char C7 { get { return 'a'; } }
		char? C8 { get { return 'a'; } }
		char? C9 { get { return null; } }
		DateTime C10 { get { return default(DateTime); } }
		int C11 { get { return -1; } }
		int C12 { get { return unchecked((int)0x80000000); } }
		long C13 { get { return 0xffffffffff; } }
		double C14 { get { return 0.5; } }
		float C15 { get { return 0.5f; } }
		sbyte C16 { get { return sbyte.MaxValue; } }
		byte C17 { get { return byte.MaxValue; } }
		short C18 { get { return short.MaxValue; } }
		ushort C19 { get { return ushort.MaxValue; } }
		uint C20 { get { return uint.MaxValue; } }
		ulong C21 { get { return ulong.MaxValue; } }
		long? C22 { get { return 0xffffffffff; } }
		double? C23 { get { return 0.5; } }
		float? C24 { get { return 0.5f; } }
		sbyte? C25 { get { return sbyte.MaxValue; } }
		byte? C26 { get { return byte.MaxValue; } }
		short? C27 { get { return short.MaxValue; } }
		ushort? C28 { get { return ushort.MaxValue; } }
		uint? C29 { get { return uint.MaxValue; } }
		ulong? C30 { get { return ulong.MaxValue; } }
		Guid C31 { get { return default(Guid); } }
		Guid? C32 { get { return default(Guid); } }
		bool C33 { get { return true; } }
		bool? C34 { get { return false; } }
		bool? C35 { get { return null; } }
		decimal C36 { get { return 1.1M; } }
		decimal? C37 { get { return 1.1M; } }
		decimal? C38 { get { return null; } }
		DateTime D39 { get { return SomeDateTime(); } }
		DateTime D40 { get { return new DateTime(2010, 1, 1, 1, 1, 1, DateTimeKind.Utc); } } // TODO
		Guid D41 { get { return new Guid("3350c137-a6c7-4e3f-aa4d-182f068899c5"); } } // TODO
		DelegateBuilderTest.B<int> D42 { get { return new DelegateBuilderTest.B<int>(); } }
		double C43 { get { return 0; } }
		float C44 { get { return 0; } }
		sbyte C45 { get { return -2; } }
		short C46 { get { return -2; } }
		int C47 { get { return -2; } }
		long C48 { get { return -2; } }
		char C49 { get { return '\u0000'; } }
		string C50 { get { return "\r\n"; } }
		string C51 { get { return "\uFF72\u3093\u4E47 \u4E02\u4E47c\u5C3A\u4E47\uFF72 \uFF89\u4E02 ou\uFF72."; } }
		#endregion
		static DateTime SomeDateTime() { return new DateTime(2010, 1, 1, 1, 1, 1, DateTimeKind.Utc); }
		#region inner types
		class OneConstant
		{
			public int A { get { return 0; } }
		}
		class OneConstantThen
		{
			public int A { get { return 0; } }
			public int B { get; set; }
		}
		class OneConstantVirtual
		{
			public virtual int A { get { return 0; } }
		}
		sealed class OneConstantSealed : OneConstantVirtual
		{
		}
		#endregion

		static bool ToleratedFailure(Type type, string newtonsoft)
		{
			var underlying = Nullable.GetUnderlyingType(type) ?? type;

			return newtonsoft != null && underlying == typeof(decimal);
		}

		[Fact]
		public void ApproveConstants()
		{
			var methods = this.GetType().GetMethods(BindingFlags.Instance | BindingFlags.NonPublic);

			var sb = new StringBuilder();
			int failedEqualCount = 0;

			foreach (var mi in methods)
			{
				if (!mi.IsSpecialName || !mi.Name.StartsWith("get_")) continue;

				var shouldBeConstant = mi.Name[4] == 'C';
				var actual = mi.Invoke(this, null);
				var nJSON = Newtonsoft.Json.JsonConvert.SerializeObject(actual);
				var constant = ConstantMethods.TryGetJson(mi, this.GetType());

				var passed = shouldBeConstant ?
					String.Equals(nJSON, constant, StringComparison.Ordinal) :
					constant == null;

				sb.AppendLine("# " + mi.Name.Substring(4) + " " + SchemaTest.HumanName(mi.ReturnType));
				sb.AppendLine();
				sb.AppendLine("Expected constant: " + shouldBeConstant);
				if (!passed)
				{
					if (ToleratedFailure(mi.ReturnType, nJSON))
					{
						sb.AppendLine("** SKIPPED ** - fix this");
					}
					else
					{
						failedEqualCount++;
						sb.AppendLine("** FAILED **");
					}
				}

				sb.AppendLine();
				sb.AppendLine("## Newtonsoft");
				Util.Hex.Dump(sb, Encoding.UTF8.GetBytes(nJSON));
				sb.AppendLine();

				sb.AppendLine();
				sb.AppendLine("## Cameronism.Json");
				if (constant != null)
				{
					Util.Hex.Dump(sb, Encoding.UTF8.GetBytes(constant));
				}
				else
				{
					sb.AppendLine();
					sb.AppendLine("`null`");
				}
				sb.AppendLine();
				sb.AppendLine();
				sb.AppendLine();
			}

			ApprovalTests.Approvals.Verify(sb.ToString());
			Assert.True(failedEqualCount == 0, "Look at the approval, " + failedEqualCount + " failed");
		}

		[Fact]
		public void ConstantObjects()
		{
			var sb = new StringBuilder();
			int failedCount = 0;

			TrySerializer(sb, new OneConstant());
			TrySerializer(sb, new OneConstantThen());
			TrySerializer(sb, new OneConstantVirtual());
			TrySerializer(sb, new OneConstantSealed());

			ApprovalTests.Approvals.Verify(sb.ToString());
			Assert.True(failedCount == 0, "Look at the approval, " + failedCount + " failed");
		}

		unsafe static int TrySerializer<T>(StringBuilder sb, params T[] samples)
		{
			int failures = 0;

			sb.AppendLine();
			sb.AppendLine("# " + SchemaTest.HumanName(typeof(T)));
			sb.AppendLine();

			var emit = DelegateBuilder.CreateStream<T>(Schema.Reflect(typeof(T)));
			string instructions;
			var del = (Serializer.WriteToStream<T>)emit.CreateDelegate(typeof(Serializer.WriteToStream<T>), out instructions);
			sb.AppendLine(emit.Instructions());
			sb.AppendLine();

			var ms = new MemoryStream();
			var buffer = new byte[64];

			for (int i = 0; i < samples.Length; i++)
			{
				ms.Position = 0;
				fixed (byte* ptr = buffer)
				{
					del.Invoke(ref samples[i], ptr, ms, buffer);
				}

				var newt = DelegateBuilderTest.GetNewtonsoft(samples[i]);
				var mine = ms.GetBuffer().Take((int)ms.Position);

				sb.AppendLine();
				sb.AppendLine("## Newtonsoft");
				Util.Hex.Dump(sb, newt);
				sb.AppendLine();

				sb.AppendLine();
				sb.AppendLine("## Cameronism.Json");
				Util.Hex.Dump(sb, mine);
				sb.AppendLine();
				sb.AppendLine();

				bool equal = newt.SequenceEqual(mine);
				sb.AppendLine("Equal: " + equal);
				sb.AppendLine();

				if (!equal) failures++;
			}

			return failures;
		}
	}
}
