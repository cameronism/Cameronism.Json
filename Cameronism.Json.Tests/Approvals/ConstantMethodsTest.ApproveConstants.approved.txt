# C2 System.String

Expected constant: True

## Newtonsoft


0000   6e 75 6c 6c                                         null

## Cameronism.Json


0000   6e 75 6c 6c                                         null


# C3 System.String

Expected constant: True

## Newtonsoft


0000   22 61 22                                            "a"

## Cameronism.Json


0000   22 61 22                                            "a"


# C4 System.Int32

Expected constant: True

## Newtonsoft


0000   31                                                  1

## Cameronism.Json


0000   31                                                  1


# C5 System.Nullable<System.Int32>

Expected constant: True

## Newtonsoft


0000   31                                                  1

## Cameronism.Json


0000   31                                                  1


# C6 System.Nullable<System.Int32>

Expected constant: True

## Newtonsoft


0000   6e 75 6c 6c                                         null

## Cameronism.Json


0000   6e 75 6c 6c                                         null


# C7 System.Char

Expected constant: True

## Newtonsoft


0000   22 61 22                                            "a"

## Cameronism.Json


0000   22 61 22                                            "a"


# C8 System.Nullable<System.Char>

Expected constant: True

## Newtonsoft


0000   22 61 22                                            "a"

## Cameronism.Json


0000   22 61 22                                            "a"


# C9 System.Nullable<System.Char>

Expected constant: True

## Newtonsoft


0000   6e 75 6c 6c                                         null

## Cameronism.Json


0000   6e 75 6c 6c                                         null


# C10 System.DateTime

Expected constant: True

## Newtonsoft


0000   22 30 30 30 31 2d 30 31   2d 30 31 54 30 30 3a 30   "0001-01 -01T00:0
0010   30 3a 30 30 22                                      0:00"

## Cameronism.Json


0000   22 30 30 30 31 2d 30 31   2d 30 31 54 30 30 3a 30   "0001-01 -01T00:0
0010   30 3a 30 30 22                                      0:00"


# C11 System.Int32

Expected constant: True

## Newtonsoft


0000   2d 31                                               -1

## Cameronism.Json


0000   2d 31                                               -1


# C12 System.Int32

Expected constant: True

## Newtonsoft


0000   2d 32 31 34 37 34 38 33   36 34 38                  -2147483 648

## Cameronism.Json


0000   2d 32 31 34 37 34 38 33   36 34 38                  -2147483 648


# C13 System.Int64

Expected constant: True

## Newtonsoft


0000   31 30 39 39 35 31 31 36   32 37 37 37 35            10995116 27775

## Cameronism.Json


0000   31 30 39 39 35 31 31 36   32 37 37 37 35            10995116 27775


# C14 System.Double

Expected constant: True

## Newtonsoft


0000   30 2e 35                                            0.5

## Cameronism.Json


0000   30 2e 35                                            0.5


# C15 System.Single

Expected constant: True

## Newtonsoft


0000   30 2e 35                                            0.5

## Cameronism.Json


0000   30 2e 35                                            0.5


# C16 System.SByte

Expected constant: True

## Newtonsoft


0000   31 32 37                                            127

## Cameronism.Json


0000   31 32 37                                            127


# C17 System.Byte

Expected constant: True

## Newtonsoft


0000   32 35 35                                            255

## Cameronism.Json


0000   32 35 35                                            255


# C18 System.Int16

Expected constant: True

## Newtonsoft


0000   33 32 37 36 37                                      32767

## Cameronism.Json


0000   33 32 37 36 37                                      32767


# C19 System.UInt16

Expected constant: True

## Newtonsoft


0000   36 35 35 33 35                                      65535

## Cameronism.Json


0000   36 35 35 33 35                                      65535


# C20 System.UInt32

Expected constant: True

## Newtonsoft


0000   34 32 39 34 39 36 37 32   39 35                     42949672 95

## Cameronism.Json


0000   34 32 39 34 39 36 37 32   39 35                     42949672 95


# C21 System.UInt64

Expected constant: True

## Newtonsoft


0000   31 38 34 34 36 37 34 34   30 37 33 37 30 39 35 35   18446744 07370955
0010   31 36 31 35                                         1615

## Cameronism.Json


0000   31 38 34 34 36 37 34 34   30 37 33 37 30 39 35 35   18446744 07370955
0010   31 36 31 35                                         1615


# C22 System.Nullable<System.Int64>

Expected constant: True

## Newtonsoft


0000   31 30 39 39 35 31 31 36   32 37 37 37 35            10995116 27775

## Cameronism.Json


0000   31 30 39 39 35 31 31 36   32 37 37 37 35            10995116 27775


# C23 System.Nullable<System.Double>

Expected constant: True

## Newtonsoft


0000   30 2e 35                                            0.5

## Cameronism.Json


0000   30 2e 35                                            0.5


# C24 System.Nullable<System.Single>

Expected constant: True

## Newtonsoft


0000   30 2e 35                                            0.5

## Cameronism.Json


0000   30 2e 35                                            0.5


# C25 System.Nullable<System.SByte>

Expected constant: True

## Newtonsoft


0000   31 32 37                                            127

## Cameronism.Json


0000   31 32 37                                            127


# C26 System.Nullable<System.Byte>

Expected constant: True

## Newtonsoft


0000   32 35 35                                            255

## Cameronism.Json


0000   32 35 35                                            255


# C27 System.Nullable<System.Int16>

Expected constant: True

## Newtonsoft


0000   33 32 37 36 37                                      32767

## Cameronism.Json


0000   33 32 37 36 37                                      32767


# C28 System.Nullable<System.UInt16>

Expected constant: True

## Newtonsoft


0000   36 35 35 33 35                                      65535

## Cameronism.Json


0000   36 35 35 33 35                                      65535


# C29 System.Nullable<System.UInt32>

Expected constant: True

## Newtonsoft


0000   34 32 39 34 39 36 37 32   39 35                     42949672 95

## Cameronism.Json


0000   34 32 39 34 39 36 37 32   39 35                     42949672 95


# C30 System.Nullable<System.UInt64>

Expected constant: True

## Newtonsoft


0000   31 38 34 34 36 37 34 34   30 37 33 37 30 39 35 35   18446744 07370955
0010   31 36 31 35                                         1615

## Cameronism.Json


0000   31 38 34 34 36 37 34 34   30 37 33 37 30 39 35 35   18446744 07370955
0010   31 36 31 35                                         1615


# C31 System.Guid

Expected constant: True

## Newtonsoft


0000   22 30 30 30 30 30 30 30   30 2d 30 30 30 30 2d 30   "0000000 0-0000-0
0010   30 30 30 2d 30 30 30 30   2d 30 30 30 30 30 30 30   000-0000 -0000000
0020   30 30 30 30 30 22                                   00000"

## Cameronism.Json


0000   22 30 30 30 30 30 30 30   30 2d 30 30 30 30 2d 30   "0000000 0-0000-0
0010   30 30 30 2d 30 30 30 30   2d 30 30 30 30 30 30 30   000-0000 -0000000
0020   30 30 30 30 30 22                                   00000"


# C32 System.Nullable<System.Guid>

Expected constant: True

## Newtonsoft


0000   22 30 30 30 30 30 30 30   30 2d 30 30 30 30 2d 30   "0000000 0-0000-0
0010   30 30 30 2d 30 30 30 30   2d 30 30 30 30 30 30 30   000-0000 -0000000
0020   30 30 30 30 30 22                                   00000"

## Cameronism.Json


0000   22 30 30 30 30 30 30 30   30 2d 30 30 30 30 2d 30   "0000000 0-0000-0
0010   30 30 30 2d 30 30 30 30   2d 30 30 30 30 30 30 30   000-0000 -0000000
0020   30 30 30 30 30 22                                   00000"


# C33 System.Boolean

Expected constant: True

## Newtonsoft


0000   74 72 75 65                                         true

## Cameronism.Json


0000   74 72 75 65                                         true


# C34 System.Nullable<System.Boolean>

Expected constant: True

## Newtonsoft


0000   66 61 6c 73 65                                      false

## Cameronism.Json


0000   66 61 6c 73 65                                      false


# C35 System.Nullable<System.Boolean>

Expected constant: True

## Newtonsoft


0000   6e 75 6c 6c                                         null

## Cameronism.Json


0000   6e 75 6c 6c                                         null


# C36 System.Decimal

Expected constant: True
** SKIPPED ** - fix this

## Newtonsoft


0000   31 2e 31                                            1.1

## Cameronism.Json

`null`



# C37 System.Nullable<System.Decimal>

Expected constant: True
** SKIPPED ** - fix this

## Newtonsoft


0000   31 2e 31                                            1.1

## Cameronism.Json

`null`



# C38 System.Nullable<System.Decimal>

Expected constant: True

## Newtonsoft


0000   6e 75 6c 6c                                         null

## Cameronism.Json


0000   6e 75 6c 6c                                         null


# D39 System.DateTime

Expected constant: False

## Newtonsoft


0000   22 32 30 31 30 2d 30 31   2d 30 31 54 30 31 3a 30   "2010-01 -01T01:0
0010   31 3a 30 31 5a 22                                   1:01Z"

## Cameronism.Json

`null`



# D40 System.DateTime

Expected constant: False

## Newtonsoft


0000   22 32 30 31 30 2d 30 31   2d 30 31 54 30 31 3a 30   "2010-01 -01T01:0
0010   31 3a 30 31 5a 22                                   1:01Z"

## Cameronism.Json

`null`



# D41 System.Guid

Expected constant: False

## Newtonsoft


0000   22 33 33 35 30 63 31 33   37 2d 61 36 63 37 2d 34   "3350c13 7-a6c7-4
0010   65 33 66 2d 61 61 34 64   2d 31 38 32 66 30 36 38   e3f-aa4d -182f068
0020   38 39 39 63 35 22                                   899c5"

## Cameronism.Json

`null`



# D42 Cameronism.Json.Tests.DelegateBuilderTest+B<System.Int32>

Expected constant: False

## Newtonsoft


0000   7b 22 69 22 3a 30 7d                                {"i":0}

## Cameronism.Json

`null`



# C43 System.Double

Expected constant: True

## Newtonsoft


0000   30 2e 30                                            0.0

## Cameronism.Json


0000   30 2e 30                                            0.0


# C44 System.Single

Expected constant: True

## Newtonsoft


0000   30 2e 30                                            0.0

## Cameronism.Json


0000   30 2e 30                                            0.0


# C45 System.SByte

Expected constant: True

## Newtonsoft


0000   2d 32                                               -2

## Cameronism.Json


0000   2d 32                                               -2


# C46 System.Int16

Expected constant: True

## Newtonsoft


0000   2d 32                                               -2

## Cameronism.Json


0000   2d 32                                               -2


# C47 System.Int32

Expected constant: True

## Newtonsoft


0000   2d 32                                               -2

## Cameronism.Json


0000   2d 32                                               -2


# C48 System.Int64

Expected constant: True

## Newtonsoft


0000   2d 32                                               -2

## Cameronism.Json


0000   2d 32                                               -2


# C49 System.Char

Expected constant: True

## Newtonsoft


0000   22 5c 75 30 30 30 30 22                             "\u0000"

## Cameronism.Json


0000   22 5c 75 30 30 30 30 22                             "\u0000"


# C50 System.String

Expected constant: True

## Newtonsoft


0000   22 5c 72 5c 6e 22                                   "\r\n"

## Cameronism.Json


0000   22 5c 72 5c 6e 22                                   "\r\n"


# C51 System.String

Expected constant: True

## Newtonsoft


0000   22 ef bd b2 e3 82 93 e4   b9 87 20 e4 b8 82 e4 b9   "....... .. .....
0010   87 63 e5 b0 ba e4 b9 87   ef bd b2 20 ef be 89 e4   .c...... ... ....
0020   b8 82 20 6f 75 ef bd b2   2e 22                     .. ou... ."

## Cameronism.Json


0000   22 ef bd b2 e3 82 93 e4   b9 87 20 e4 b8 82 e4 b9   "....... .. .....
0010   87 63 e5 b0 ba e4 b9 87   ef bd b2 20 ef be 89 e4   .c...... ... ....
0020   b8 82 20 6f 75 ef bd b2   2e 22                     .. ou... ."


